core/src/main/java/org/apache/gravitino/authorization/PermissionManager.java

```diff
@@ -373,11 +377,302 @@ User revokeRolesFromUser(String metalake, List<String> roles, String user) {
+                    .withProperties(roleEntity.properties())
+                    .withAuditInfo(auditInfo)
+                    .withSecurableObjects(updateSecurableObjects)
+                    .build();
+              });
+
+      authorizationCallbackWrapper.execute();
+
+      return updatedRole;
+    } catch (NoSuchEntityException nse) {
+      LOG.error("Failed to revoke, role {} does not exist in the metalake {}", role, metalake, nse);
+      throw new NoSuchRoleException(ROLE_DOES_NOT_EXIST_MSG, role, metalake);
+    } catch (IOException ioe) {
+      throw new RuntimeException(ioe);
+    }
+  }
+
+  private static SecurableObject updateRevokedSecurableObject(
+      String metalake,
+      String role,
+      MetadataObject object,
+      List<Privilege> privileges,
+      RoleEntity roleEntity,
+      SecurableObject targetObject,
+      AuthorizationPluginCallbackWrapper authorizationCallbackWrapper) {
+    // If the securable object exists, we remove the privileges of the
+    // securable object. Remove duplicated privileges
+    Set<Privilege> updatePrivileges = Sets.newHashSet();
+    updatePrivileges.addAll(targetObject.privileges());
+    privileges.forEach(updatePrivileges::remove);
+
+    // If the object still contains privilege, we should update the object
+    // with new privileges
+    if (!updatePrivileges.isEmpty()) {
+      SecurableObject newSecurableObject =
+          SecurableObjects.parse(
+              targetObject.fullName(), targetObject.type(), Lists.newArrayList(updatePrivileges));
+      authorizationCallbackWrapper.setCallBack(
+          () ->
+              AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                  metalake,
+                  object,
+                  authorizationPlugin -> {
+                    authorizationPlugin.onRoleUpdated(
+                        roleEntity,
+                        RoleChange.updateSecurableObject(role, targetObject, newSecurableObject));
+                  }));
+
+      return newSecurableObject;
+    } else {
+      // If the object doesn't contain any privilege, we remove this object.
+      authorizationCallbackWrapper.setCallBack(
+          () ->
+              AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                  metalake,
+                  object,
+                  authorizationPlugin -> {
+                    authorizationPlugin.onRoleUpdated(
+                        roleEntity, RoleChange.removeSecurableObject(role, targetObject));
+                  }));
+
+      return null;
+    }
+  }
+
+  private List<SecurableObject> updateSecurableObjects(
+      List<SecurableObject> securableObjects,
+      MetadataObject targetObject,
+      Function<SecurableObject, SecurableObject> objectUpdater) {
+    boolean isExist = false;
+    List<SecurableObject> updateSecurableObjects = Lists.newArrayList();
+    for (SecurableObject securableObject : securableObjects) {
```

Comment:
Do we support updating multiple securable objects once?

Reply:
If we don't support, why don't we just search the matched securable object and update it, no need to do this in a loop add a flag?

For example:

```
securableObject = get matched securableObject
if (securableObject != null) {
updater.apply(xxx)
}
```

Reply:
The method name isn't good. I should generate all the securable objects for the role. I changed to 
```
generateNewSecurableObjects
```
And I tune the logic of this method and I make it clear.

Reply:
```java
   List<SecurableObject> updateSecurableObjects = Lists.newArrayList(securableObjects);
    SecurableObject matchedSecurableObject = securableObjects.stream().filter(targetObject::equals).findFirst().orElse(null);
    if (matchedSecurableObject != null) {
      updateSecurableObjects.remove(matchedSecurableObject);
    }
```

This may be more efficient that the above.